### PR TITLE
Allow overriding root disk offering & size, and expunge old root disk while restoring a VM

### PR DESCRIPTION
**Pull request by vishesh92**
### Description

This PR allows overriding root diskoffering id & size, and expunge old root disk  while restoring VM. This is useful when you have to restore a VM with a new template but the newer template's size is smaller than the existing one.

New parameters for `restoreVirtualMachine`:
1. `diskofferingid` - Override root volume's diskoffering
2. `rootdisksize` - Override root volume's size (in GB)
3. `expunge` - Option to expunge old volumes after restore
4. `details` - Option to specify custom parameters for the selected disk offering

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### Screenshots (if appropriate):
image

### How Has This Been Tested?

Tested Manually

### CODE REVIEW

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
Code LGTM

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by gpordeus** [comment]
Hi, @vishesh92, good change!

Could you describe your tests? Have you tried it with a compute offering with `Disk offering strictness`? What about expunge permissions?

**Comment by vishesh92** [comment]
> Hi, @vishesh92, good change!
> 
> Could you describe your tests? Have you tried it with a compute offering with `Disk offering strictness`? What about expunge permissions?

I tested with compute only disk offering & with another disk offering. I didn't check/test with disk offering strictness. I have done a few changes to add the check for disk offering strictness & expunge permissions. Let me the test the changes again. I have changed the PR to draft for now.

**Comment by vishesh92** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -7642,27 +7644,52 @@ public UserVm restoreVM(RestoreVMCmd cmd) throws InsufficientCapacityException,
...
+            if (diskOffering == null) {
+                throw new InvalidParameterValueException("Cannot find disk offering with ID " + rootDiskOfferingId);
+            }
+            DataCenterVO zone = dataCenterDao.findById(vm.getDataCenterId());
+            _accountMgr.checkAccess(caller, diskOffering, zone);
+            ServiceOfferingVO serviceOffering = serviceOfferingDao.findById(vm.getServiceOfferingId());
+            if (serviceOffering.getDiskOfferingStrictness() && !serviceOffering.getDiskOfferingId().equals(rootDiskOfferingId)) {
+                throw new InvalidParameterValueException("VM's service offering has a strict disk offering requirement, and the specified disk offering does not match");
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:7675]: can this go in a new method?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -7797,15 +7826,22 @@ public Pair<UserVmVO, Volume> doInTransaction(final TransactionStatus status) th
...
                             }
                         }
+                        if (diskOffering != null) {
+                            resizedVolume.setDiskOfferingId(diskOffering.getId());
+                            resizedVolume.setSize(diskOffering.getDiskSize());
+                        }
+                        if (rootDiskSize != null) {
+                            resizedVolume.setSize(rootDiskSize);
+                        }
+                        _volsDao.update(resizedVolume.getId(), resizedVolume);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:7844]: can these bits go in new methods?

**Review by DaanHoogland — COMMENTED** [review]
code looks good, needs testing

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by vishesh92** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@blueorangutan test

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested

**Pull request merged**